File: libs/shared-types/validators/python/featurecollection_validator.py
"""
Manual validators for FeatureCollection
These validators work with the generated types and provide additional validation logic
"""

from typing import Any, Dict, List, Union, Optional, Tuple
from datetime import datetime
import sys
import os

# Add the derived module to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'derived', 'python'))

# Import validator functions from other modules
try:
    from .track_validator import validate_track_feature
    from .point_validator import validate_point_feature  
    from .annotation_validator import validate_annotation_feature
except ImportError:
    # Handle relative import issues during testing
    import track_validator
    import point_validator
    import annotation_validator
    validate_track_feature = track_validator.validate_track_feature
    validate_point_feature = point_validator.validate_point_feature
    validate_annotation_feature = annotation_validator.validate_annotation_feature

try:
    from featurecollection import DebriefFeatureCollection
except ImportError:
    # If import fails, define a minimal type for validation
    class DebriefFeatureCollection:
        pass


def validate_bbox(bbox: List[float]) -> bool:
    """
    Validates bounding box format
    """
    # Must be either 2D bbox [minX, minY, maxX, maxY] or 3D bbox [minX, minY, minZ, maxX, maxY, maxZ]
    if len(bbox) not in [4, 6]:
        return False
    
    # Check all values are numbers and finite
    if not all(isinstance(val, (int, float)) and not (val != val or val == float('inf') or val == float('-inf')) for val in bbox):
        return False
    
    # For 2D bbox: minX <= maxX and minY <= maxY
    if len(bbox) == 4:
        min_x, min_y, max_x, max_y = bbox
        return min_x <= max_x and min_y <= max_y
    
    # For 3D bbox: minX <= maxX, minY <= maxY, minZ <= maxZ
    if len(bbox) == 6:
        min_x, min_y, min_z, max_x, max_y, max_z = bbox
        return min_x <= max_x and min_y <= max_y and min_z <= max_z
    
    return False


def validate_feature_collection(collection: Any) -> bool:
    """
    Validates that feature collection has required properties and valid structure
    """
    if not isinstance(collection, dict):
        return False
    
    # Check basic GeoJSON FeatureCollection structure
    if collection.get('type') != 'FeatureCollection':
        return False
    
    features = collection.get('features')
    if not isinstance(features, list):
        return False
    
    # Validate bbox if present
    bbox = collection.get('bbox')
    if bbox is not None and not validate_bbox(bbox):
        return False
    
    # Validate each feature has required id property
    for feature in features:
        feature_id = feature.get('id')
        if feature_id is None:
            return False  # id is required for all features
    
    return True


def classify_feature(feature: Dict[str, Any]) -> str:
    """
    Determines the type of feature based on geometry and properties
    """
    if not isinstance(feature, dict) or 'geometry' not in feature:
        return 'unknown'
    
    geometry = feature.get('geometry', {})
    geometry_type = geometry.get('type')
    properties = feature.get('properties', {})
    
    # Track features: LineString or MultiLineString with optional timestamps
    if geometry_type in ['LineString', 'MultiLineString']:
        # If it has annotationType, it's an annotation, not a track
        if properties.get('annotationType'):
            return 'annotation'
        return 'track'
    
    # Point features: Point geometry without annotationType
    if geometry_type == 'Point':
        if properties.get('annotationType'):
            return 'annotation'
        return 'point'
    
    # All other geometries are annotations
    if geometry_type in ['Polygon', 'MultiPoint', 'MultiPolygon']:
        return 'annotation'
    
    return 'unknown'


def validate_feature_by_type(feature: Any) -> bool:
    """
    Validates individual feature based on its type
    """
    feature_type = classify_feature(feature)
    
    if feature_type == 'track':
        return validate_track_feature(feature)
    elif feature_type == 'point':
        return validate_point_feature(feature)
    elif feature_type == 'annotation':
        return validate_annotation_feature(feature)
    else:
        return False


def is_valid_date(date_value: Any) -> bool:
    """
    Validates that date string or datetime object is valid
    """
    if isinstance(date_value, datetime):
        return True
    
    if isinstance(date_value, str):
        try:
            datetime.fromisoformat(date_value.replace('Z', '+00:00'))
            return 'T' in date_value  # Expect ISO format
        except ValueError:
            return False
    
    return False


def validate_feature_collection_properties(properties: Any) -> bool:
    """
    Validates feature collection properties
    """
    if not properties or not isinstance(properties, dict):
        return True  # properties are optional
    
    # Validate time properties if present
    created = properties.get('created')
    if created and not is_valid_date(created):
        return False
    
    modified = properties.get('modified')
    if modified and not is_valid_date(modified):
        return False
    
    # If both created and modified are present, created should be <= modified
    if created and modified:
        try:
            if isinstance(created, str):
                created_dt = datetime.fromisoformat(created.replace('Z', '+00:00'))
            else:
                created_dt = created
                
            if isinstance(modified, str):
                modified_dt = datetime.fromisoformat(modified.replace('Z', '+00:00'))
            else:
                modified_dt = modified
                
            if created_dt > modified_dt:
                return False
        except (ValueError, TypeError):
            return False
    
    return True


def get_feature_counts(collection: Dict[str, Any]) -> Dict[str, int]:
    """
    Gets feature counts by type
    """
    features = collection.get('features', [])
    counts = {
        'tracks': 0,
        'points': 0,
        'annotations': 0,
        'unknown': 0,
        'total': len(features)
    }
    
    for feature in features:
        feature_type = classify_feature(feature)
        if feature_type in counts:
            counts[feature_type + 's'] += 1
        else:
            counts['unknown'] += 1
    
    return counts


def validate_feature_collection_comprehensive(collection: Any) -> Dict[str, Union[bool, List[str], Dict[str, int]]]:
    """
    Comprehensive feature collection validation
    """
    errors = []
    
    # Basic validation
    if not validate_feature_collection(collection):
        errors.append('Invalid feature collection structure')
        return {'is_valid': False, 'errors': errors}
    
    # Properties validation
    properties = collection.get('properties')
    if not validate_feature_collection_properties(properties):
        errors.append('Invalid feature collection properties')
    
    # Validate bbox if present
    bbox = collection.get('bbox')
    if bbox and not validate_bbox(bbox):
        errors.append('Invalid bounding box format')
    
    # Validate each feature
    features = collection.get('features', [])
    for index, feature in enumerate(features):
        if not validate_feature_by_type(feature):
            errors.append(f'Feature at index {index} is invalid')
    
    # Get feature counts for analysis
    feature_counts = get_feature_counts(collection) if len(errors) == 0 else None
    
    result = {
        'is_valid': len(errors) == 0,
        'errors': errors
    }
    
    if feature_counts:
        result['feature_counts'] = feature_counts
    
    return result

File: libs/shared-types/src/validators/featurecollection-validator.ts
/**
 * Manual validators for FeatureCollection
 * These validators work with the generated types and provide additional validation logic
 */

import { DebriefFeatureCollection } from '../types/features/debrief_feature_collection';
import { validateTrackFeature } from './track-validator';
import { validatePointFeature } from './point-validator';
import { validateAnnotationFeature, validateAnnotationType, validateColorFormat } from './annotation-validator';

/**
 * Type-safe helper to check if a value is a non-null object
 */
function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object';
}

/**
 * Type-safe helper to access object properties
 */
function getObjectProperty(obj: Record<string, unknown>, key: string): unknown {
  return obj[key];
}

/**
 * Validates bounding box format
 */
export function validateBbox(bbox: number[]): boolean {
  // Must be either 2D bbox [minX, minY, maxX, maxY] or 3D bbox [minX, minY, minZ, maxX, maxY, maxZ]
  if (bbox.length !== 4 && bbox.length !== 6) {
    return false;
  }
  
  // Check all values are numbers
  if (!bbox.every(val => typeof val === 'number' && isFinite(val))) {
    return false;
  }
  
  // For 2D bbox: minX <= maxX and minY <= maxY
  if (bbox.length === 4) {
    const [minX, minY, maxX, maxY] = bbox;
    return minX <= maxX && minY <= maxY;
  }
  
  // For 3D bbox: minX <= maxX, minY <= maxY, minZ <= maxZ
  if (bbox.length === 6) {
    const [minX, minY, minZ, maxX, maxY, maxZ] = bbox;
    return minX <= maxX && minY <= maxY && minZ <= maxZ;
  }
  
  return false;
}

/**
 * Validates that feature collection has required properties and valid structure
 */
export function validateFeatureCollection(collection: unknown): collection is DebriefFeatureCollection {
  if (!isObject(collection)) {
    return false;
  }
  
  // Check basic GeoJSON FeatureCollection structure
  if (getObjectProperty(collection, 'type') !== 'FeatureCollection') {
    return false;
  }
  
  const features = getObjectProperty(collection, 'features');
  if (!Array.isArray(features)) {
    return false;
  }
  
  // Validate bbox if present
  const bbox = getObjectProperty(collection, 'bbox');
  if (bbox !== undefined && (!Array.isArray(bbox) || !validateBbox(bbox))) {
    return false;
  }
  
  // Validate each feature has required id property
  for (const feature of features) {
    if (!isObject(feature)) {
      return false;
    }
    const id = getObjectProperty(feature, 'id');
    if (id === null || id === undefined) {
      return false; // id is required for all features
    }
  }
  
  return true;
}

/**
 * Determines the type of feature based on geometry and properties
 */
export function classifyFeature(feature: unknown): 'track' | 'point' | 'annotation' | 'unknown' {
  if (!isObject(feature)) {
    return 'unknown';
  }
  
  const geometry = getObjectProperty(feature, 'geometry');
  if (!isObject(geometry)) {
    return 'unknown';
  }
  
  const properties = getObjectProperty(feature, 'properties');
  const propertiesObj = isObject(properties) ? properties : {};
  
  // Skip validation for unsupported dataType values
  const dataType = getObjectProperty(propertiesObj, 'dataType');
  if (dataType === 'buoyfield' || dataType === 'backdrop' || dataType === 'metadata') {
    return 'unknown';
  }
  
  const geometryType = getObjectProperty(geometry, 'type');
  
  // Use dataType for classification if available
  if (typeof dataType === 'string') {
    switch (dataType) {
      case 'track':
        return 'track';
      case 'reference-point':
        return 'point';
      case 'zone':
        return 'annotation';
      default:
        // Invalid dataType values should still be validated based on geometry
        // Only 'buoyfield' and 'backdrop' should be skipped (handled above)
        break;
    }
  }
  
  // Fallback to geometry-based classification for backward compatibility
  // Track features: LineString or MultiLineString with optional timestamps
  if (geometryType === 'LineString' || geometryType === 'MultiLineString') {
    // If it has annotationType, it's an annotation, not a track
    const annotationType = getObjectProperty(propertiesObj, 'annotationType');
    if (annotationType !== undefined) {
      return 'annotation';
    }
    return 'track';
  }
  
  // Point features: Point geometry without annotationType
  if (geometryType === 'Point') {
    const annotationType = getObjectProperty(propertiesObj, 'annotationType');
    if (annotationType !== undefined) {
      return 'annotation';
    }
    return 'point';
  }
  
  // All other geometries are annotations
  if (typeof geometryType === 'string' && ['Polygon', 'MultiPoint', 'MultiPolygon'].indexOf(geometryType) !== -1) {
    return 'annotation';
  }
  
  return 'unknown';
}

/**
 * Validates individual feature based on its type
 */
export function validateFeatureByType(feature: unknown): boolean {
  const featureType = classifyFeature(feature);
  
  switch (featureType) {
    case 'track':
      return validateTrackFeature(feature);
    case 'point':
      return validatePointFeature(feature);
    case 'annotation':
      return validateAnnotationFeature(feature);
    case 'unknown':
      // Skip validation for unknown types (including buoyfield and backdrop)
      return true;
    default:
      return false;
  }
}

/**
 * Validates that date string or Date object is valid
 */
export function isValidDate(dateValue: unknown): boolean {
  if (dateValue instanceof Date) {
    return !isNaN(dateValue.getTime());
  }
  
  if (typeof dateValue === 'string') {
    const parsed = new Date(dateValue);
    return !isNaN(parsed.getTime()) && dateValue.indexOf('T') !== -1; // Expect ISO format
  }
  
  return false;
}

/**
 * Validates feature collection properties
 */
export function validateFeatureCollectionProperties(properties: unknown): boolean {
  if (!isObject(properties)) {
    return true; // properties are optional
  }
  
  // Validate time properties if present
  const created = getObjectProperty(properties, 'created');
  if (created !== undefined && !isValidDate(created)) {
    return false;
  }
  
  const modified = getObjectProperty(properties, 'modified');
  if (modified !== undefined && !isValidDate(modified)) {
    return false;
  }
  
  // If both created and modified are present, created should be <= modified
  if (created !== undefined && modified !== undefined) {
    const createdDate = new Date(created as string | Date);
    const modifiedDate = new Date(modified as string | Date);
    if (createdDate > modifiedDate) {
      return false;
    }
  }
  
  return true;
}

/**
 * Gets feature counts by type
 */
export function getFeatureCounts(collection: DebriefFeatureCollection): {
  tracks: number;
  points: number;
  annotations: number;
  unknown: number;
  total: number;
} {
  const counts = {
    tracks: 0,
    points: 0,
    annotations: 0,
    unknown: 0,
    total: collection.features.length
  };
  
  collection.features.forEach(feature => {
    const type = classifyFeature(feature);
    switch (type) {
      case 'track':
        counts.tracks++;
        break;
      case 'point':
        counts.points++;
        break;
      case 'annotation':
        counts.annotations++;
        break;
      default:
        counts.unknown++;
        break;
    }
  });
  
  return counts;
}

/**
 * Enhanced validation result with detailed feature information
 */
export interface FeatureValidationResult {
  isValid: boolean;
  errors: string[];
  featureCounts?: ReturnType<typeof getFeatureCounts>;
  featureErrors?: FeatureError[];
}

/**
 * Detailed error information for a specific feature
 */
export interface FeatureError {
  index: number;
  featureId?: string | number;
  featureType: 'track' | 'point' | 'annotation' | 'unknown';
  errors: string[];
}

/**
 * Get feature identifier for error reporting
 */
function getFeatureIdentifier(feature: unknown, index: number): string {
  if (!isObject(feature)) {
    return `index ${index} (no id)`;
  }
  const id = getObjectProperty(feature, 'id');
  if (id !== undefined && id !== null) {
    return `index ${index} (id: "${id}")`;
  }
  return `index ${index} (no id)`;
}

/**
 * Validates individual feature with detailed error reporting
 */
export function validateFeatureDetailed(feature: unknown, index: number): FeatureError | null {
  const featureType = classifyFeature(feature);
  const errors: string[] = [];
  
  // Basic structure validation
  if (!isObject(feature)) {
    errors.push('Feature is not an object');
  } else {
    const featureType = getObjectProperty(feature, 'type');
    if (featureType !== 'Feature') {
      errors.push(`Invalid feature type: "${featureType}" (expected "Feature")`);
    }
    
    const id = getObjectProperty(feature, 'id');
    if (id === null || id === undefined) {
      errors.push('Missing required "id" property');
    }
    
    const geometry = getObjectProperty(feature, 'geometry');
    if (!isObject(geometry)) {
      errors.push('Missing or invalid geometry object');
    }
    
    const properties = getObjectProperty(feature, 'properties');
    if (!isObject(properties)) {
      errors.push('Missing or invalid properties object');
    }
  }
  
  // Type-specific validation
  if (errors.length === 0 && isObject(feature)) {
    let isValidByType = false;
    const geometry = getObjectProperty(feature, 'geometry');
    const properties = getObjectProperty(feature, 'properties');
    
    switch (featureType) {
      case 'track':
        isValidByType = validateTrackFeature(feature);
        if (!isValidByType) {
          errors.push('Failed track-specific validation');
          // Add more specific track validation errors
          if (isObject(geometry)) {
            const geometryType = getObjectProperty(geometry, 'type');
            if (typeof geometryType === 'string' && ['LineString', 'MultiLineString'].indexOf(geometryType) === -1) {
              errors.push(`Track must have LineString or MultiLineString geometry, got "${geometryType}"`);
            }
          }
          if (isObject(properties)) {
            const dataType = getObjectProperty(properties, 'dataType');
            if (!dataType) {
              errors.push('Track features must have dataType: "track" in properties');
            }
          }
        }
        break;
        
      case 'point':
        isValidByType = validatePointFeature(feature);
        if (!isValidByType) {
          errors.push('Failed point-specific validation');
          // Add more specific point validation errors
          if (isObject(geometry)) {
            const geometryType = getObjectProperty(geometry, 'type');
            if (geometryType !== 'Point') {
              errors.push(`Point feature must have Point geometry, got "${geometryType}"`);
            }
          }
          if (isObject(properties)) {
            const dataType = getObjectProperty(properties, 'dataType');
            if (!dataType) {
              errors.push('Point features must have dataType: "reference-point" in properties');
            }
            const time = getObjectProperty(properties, 'time');
            const timeStart = getObjectProperty(properties, 'timeStart');
            const timeEnd = getObjectProperty(properties, 'timeEnd');
            if (time && (timeStart || timeEnd)) {
              errors.push('Cannot have both single time and time range (timeStart/timeEnd)');
            }
            if ((timeStart && !timeEnd) || (!timeStart && timeEnd)) {
              errors.push('Time range requires both timeStart and timeEnd');
            }
          }
        }
        break;
        
      case 'annotation':
        isValidByType = validateAnnotationFeature(feature);
        if (!isValidByType) {
          errors.push('Failed annotation-specific validation');
          // Add more specific annotation validation errors
          if (isObject(properties)) {
            const dataType = getObjectProperty(properties, 'dataType');
            if (!dataType) {
              errors.push('Annotation features must have dataType: "zone" in properties');
            }
            const annotationType = getObjectProperty(properties, 'annotationType');
            if (typeof annotationType === 'string' && !validateAnnotationType(annotationType)) {
              errors.push(`Invalid annotation type: "${annotationType}"`);
            }
            const color = getObjectProperty(properties, 'color');
            if (typeof color === 'string' && !validateColorFormat(color)) {
              errors.push(`Invalid color format: "${color}" (expected #RRGGBB)`);
            }
          }
        }
        break;
        
      default:
        // Don't add errors for unknown types - they are valid but not validated
        // This includes buoyfield and backdrop dataTypes
        break;
    }
  }
  
  if (errors.length === 0) {
    return null;
  }
  
  const featureId = isObject(feature) ? getObjectProperty(feature, 'id') : undefined;
  return {
    index,
    featureId: typeof featureId === 'string' || typeof featureId === 'number' ? featureId : undefined,
    featureType,
    errors
  };
}

/**
 * Comprehensive feature collection validation with enhanced error reporting
 */
export function validateFeatureCollectionComprehensive(collection: unknown): FeatureValidationResult {
  const errors: string[] = [];
  const featureErrors: FeatureError[] = [];
  
  // Basic validation
  if (!validateFeatureCollection(collection)) {
    errors.push('Invalid feature collection structure');
    return { isValid: false, errors };
  }
  
  const collectionObj = collection as DebriefFeatureCollection;
  
  // Properties validation
  if (!validateFeatureCollectionProperties(collectionObj.properties)) {
    errors.push('Invalid feature collection properties');
  }
  
  // Validate bbox if present
  if (collectionObj.bbox && !validateBbox(collectionObj.bbox)) {
    errors.push('Invalid bounding box format');
  }
  
  // Validate each feature with detailed error reporting
  collectionObj.features.forEach((feature: unknown, index: number) => {
    const featureError = validateFeatureDetailed(feature, index);
    if (featureError) {
      featureErrors.push(featureError);
      // Add summary error message
      const identifier = getFeatureIdentifier(feature, index);
      errors.push(`Feature at ${identifier} is invalid (${featureError.errors.length} error${featureError.errors.length > 1 ? 's' : ''})`);
    }
  });
  
  // Get feature counts for analysis
  const featureCounts = errors.length === 0 ? getFeatureCounts(collectionObj) : undefined;
  
  return {
    isValid: errors.length === 0,
    errors,
    featureCounts,
    featureErrors: featureErrors.length > 0 ? featureErrors : undefined
  };
}